boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,211 @@
+    // Invert watchdog bit state
+    if (e_fuse->watch_dog_state)
+    {
+        SET_BIT(command, WATCH_DOG_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, WATCH_DOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->watch_dog_state = !e_fuse->watch_dog_state;
+    // Compute parity
+    set_parity_bit = Io_Efuse_Set_Parity_Bit(command);
+    // Set parity value
+    if (set_parity_bit)
+    {
+        SET_BIT(command, PARITY_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, PARITY_BIT);
+    }
+    status = Io_Efuse_WriteToEfuse(
+        &command, e_fuse->chip_select.GPIO_Port, e_fuse->chip_select.GPIO_Pin);
+    UNUSED(status);
+}
+
+uint16_t Io_Efuse_ReadReg(uint8_t register_address, struct Efuse *e_fuse)
+{
+    HAL_StatusTypeDef status;
+    uint16_t          command        = 0x0000U;
+    bool              set_parity_bit = false;
+    uint16_t          tx_data[]      = { 0x0000U };
+    uint16_t          rx_data[]      = { 0x0000U };
+
+    // Place the Status Register address into bits 10->13
+    command =
+        (uint16_t)((SI_STATR_0_ADDR & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    // Shift bit 3 of the address (SOA3: the channel number) to the 13th bit
+    command = (uint16_t)(
+        command | ((register_address & SOA3_MASK) << EFUSE_ADDR_SHIFT) |
+        (register_address & (SOA2_MASK | SOA1_MASK | SOA0_MASK)));
+    // Invert watchdog bit state
+    if (e_fuse->watch_dog_state)
+    {
+        command |= WATCH_DOG_BIT;
+    }
+    else
+    {
+        command &= ~(WATCH_DOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->watch_dog_state = !e_fuse->watch_dog_state;
+    // Compute parity
+    set_parity_bit = Io_Efuse_Set_Parity_Bit(command);
+    // Set parity value
+    if (set_parity_bit)
+    {
+        SET_BIT(command, PARITY_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, PARITY_BIT);
+    }
+    tx_data[0] = command;
+    status     = Io_Efuse_ReadFromEfuse(
+        tx_data, rx_data, e_fuse->chip_select.GPIO_Port,
+        e_fuse->chip_select.GPIO_Pin);
+    UNUSED(status);
+    // Only return register contents and clear bits 9->15
+    CLEAR_BIT(rx_data[0], ~EFUSE_SO_DATA_MASK);
+    return rx_data[0];
+}
+
+bool Io_Efuse_Set_Parity_Bit(uint16_t spi_command)
+{
+    uint8_t parity_bit;
+    // Calculate parity for bits in spi command
+    for (parity_bit = 0; spi_command > 0; spi_command >>= 1)
+    {
+        parity_bit ^= (spi_command & 1);
+    }
+    return ((parity_bit == 1) ? true : false);
+}
+
+HAL_StatusTypeDef Io_Efuse_WriteToEfuse(
+    uint16_t *    TxData,
+    GPIO_TypeDef *GPIO_Port,
+    uint16_t      GPIO_Pin)
+{
+    HAL_StatusTypeDef status;
+
+    HAL_GPIO_WritePin(GPIO_Port, GPIO_Pin, GPIO_PIN_RESET);
+    status = HAL_SPI_Transmit(efuse_spi_handle, (uint8_t *)TxData, 1U, 100U);
+    while (efuse_spi_handle->State != HAL_SPI_STATE_READY)
+    {
+    }
```

Comment:
Is this while-loop actually needed? The example SPI transmit code I found only never required a while-loop wait like this and it's worked fine for my DIM 7-seg driver too

Reply:
Ya I don't think so either, I had it in there previously during debugging. I think due to the timeout of the polling HAL_SPI calls, the peripheral will always be ready either before the timeout period or after, if a timeout occurred.

Reply:
I see it's removed now